### PR TITLE
Patch errors made for when user enters wrong modality into a structured pointer field

### PR DESCRIPTION
**Pull request by RaynorChavez**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
Generic Internal error when user 
- adds text to a video/audio pointer field in structured
- adds video link to audio pointer and vice version in structured

* **What is the new behavior (if this is a feature change)?**
Only the specific document is failed to 400, but the batch is not.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)

* **Related Python client changes** (link commit/PR here)

* **Related documentation changes** (link commit/PR here)

* **Other information**:

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/tensor_search/add_docs.py [hunk 1]
```diff
@@ -91,7 +91,7 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
                     modality = infer_modality(doc[field])
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: rename it to `inferred_modality` for clarificaiton

File: src/marqo/tensor_search/add_docs.py [hunk 2]
```diff
@@ -129,25 +129,40 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
...
+                        if is_structured_index:
+                            if modality is Modality.VIDEO and media_field_types_mapping[
+                            field] is not FieldType.VideoPointer:
+                                media_repo[doc[field]] = S2InferenceError(
+                                    f"Invalid audio file. Error processing media file {doc}, detected as video, but field type is not VideoPointer")
+                                continue
+
+                            if modality is Modality.AUDIO and media_field_types_mapping[
+                            field] is not FieldType.AudioPointer:
+                                media_repo[doc[field]] = S2InferenceError(
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: Create a new error for this. S2InferenceError is too generic. Maybe `MediaMismatchError`?  Make sure the error is caught properly so it does not break the whole batch but just returns 400 for the document.

File: src/marqo/tensor_search/add_docs.py [hunk 3]
```diff
@@ -91,7 +91,7 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
                     modality = infer_modality(doc[field])
                     if (modality == Modality.IMAGE and is_no_index) or (
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: It's hard to follow the whole logic here. Could it be simplified? Like raise the error earlier if inferred error does not match the field?

**Review by wanliAlex — CHANGES_REQUESTED** [review]

File: src/marqo/tensor_search/add_docs.py (hunk 1 above)
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: Your infer_modality function is not reliable:
1. It does not encoded the url
2. It shouldn't return text if there is an exception. What if a time-out error happens?

File: src/marqo/s2_inference/multimodal_model_load.py [hunk 4]
```diff
@@ -143,13 +145,19 @@ def infer_modality(content: Union[str, List[str], bytes]) -> Modality:
     if isinstance(content, str):
         if not validate_url(content):
             return Modality.TEXT
+        
+        # Encode the URL
+        encoded_url = quote(content, safe=':/')
```
Comment (wanliAlex) [inline src/marqo/s2_inference/multimodal_model_load.py]: Use the function here: https://github.com/marqo-ai/marqo/blob/0f09c36f2f2ab0add8cbb07ce939000fffb932d3/src/marqo/core/inference/image_download.py#L218

File: src/marqo/s2_inference/multimodal_model_load.py [hunk 5]
```diff
@@ -158,13 +166,17 @@ def infer_modality(content: Union[str, List[str], bytes]) -> Modality:
                 with fetch_content_sample(content) as sample:
```
Comment (wanliAlex) [inline src/marqo/s2_inference/multimodal_model_load.py]: we should pass the encoded_url when fetching the sample?

File: src/marqo/s2_inference/multimodal_model_load.py [hunk 6]
```diff
@@ -158,13 +166,17 @@ def infer_modality(content: Union[str, List[str], bytes]) -> Modality:
...
+                        print(f"inferred video from url")
                         return Modality.VIDEO
                     elif mime.startswith('audio/'):
+                        print(f"inferred audio from url")
                         return Modality.AUDIO
-            except Exception as e:
-                pass
+            except requests.exceptions.RequestException as e:
+                print(f"Request failed: {e}")
+                raise ValueError("Failed to fetch content from URL.")
```
Comment (wanliAlex) [inline src/marqo/s2_inference/multimodal_model_load.py]: Why do we raise a ValueError here? Will this become an internal error?

File: src/marqo/tensor_search/add_docs.py (hunk 1 above)
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: Should we catch the ValueError here?

**Review by wanliAlex — CHANGES_REQUESTED** [review]

File: src/marqo/tensor_search/add_docs.py (hunk 1 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: done

File: src/marqo/s2_inference/multimodal_model_load.py (hunk 4 above)
Comment (RaynorChavez) [reply src/marqo/s2_inference/multimodal_model_load.py]: done

File: src/marqo/s2_inference/multimodal_model_load.py (hunk 5 above)
Comment (RaynorChavez) [reply src/marqo/s2_inference/multimodal_model_load.py]: done

File: src/marqo/s2_inference/multimodal_model_load.py (hunk 6 above)
Comment (RaynorChavez) [reply src/marqo/s2_inference/multimodal_model_load.py]: revert to pass

File: src/marqo/tensor_search/add_docs.py (hunk 1 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: done

File: src/marqo/tensor_search/add_docs.py (hunk 3 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: will refactor in the next release!

File: src/marqo/tensor_search/add_docs.py (hunk 2 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: This might affect a bunch of tests/logic we have, esp in tensor_search. This will require a bit more investigation. Not sure if it fits here. I think we can leave it at `S2InferenceError` for now

File: src/marqo/tensor_search/add_docs.py (hunk 1 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: I think the pass and being caught in add_docs.py should suffice for now

File: src/marqo/tensor_search/add_docs.py [hunk 7]
```diff
@@ -88,10 +88,10 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
                 if field not in tensor_fields:
                     continue
                 if isinstance(doc[field], str) or force_download:
-                    modality = infer_modality(doc[field])
-                    if (modality == Modality.IMAGE and is_no_index) or (
-                        modality == Modality.IMAGE and is_unstructured_index) or (
-                        is_structured_index and media_field_types_mapping[field] == FieldType.ImagePointer):
+                    inferred_modality = infer_modality(doc[field])
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: Why don't we catch the requests error here and return the error message to the user?

File: src/marqo/tensor_search/add_docs.py [hunk 8]
```diff
@@ -88,10 +88,10 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
                 if field not in tensor_fields:
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py:90]: @farshidz 
There are two ways to solve the issue when users provide an incorrect modality for a structured index:
- We infer the modality and see if there is a mismatch, which is what we are doing in this PR
- We don’t infer the modality and just process it as if it is the expected content and catch the raised error

Option 1 is more effective, but we need to be confident with our inferred modality.
Option 2 is more straightforward but it is more complicated and slower because we need to download the content, process it, and catch the error.

I think option 1 is good because we are using a widely used `mime` package now. Your thoughts?

File: src/marqo/tensor_search/add_docs.py [hunk 9]
```diff
@@ -127,27 +127,42 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
...
                             media_repo[doc[field]] = processed_chunks
                         except (ffmpeg.Error, S2InferenceError) as e:
-                            logger.error(f"Error processing {modality} file: {str(e)}")
-                            media_repo[doc[field]] = S2InferenceError(f"Error processing {modality} file: {str(e)}")
-
+                            logger.error(f"Error processing {inferred_modality} file: {str(e)}")
+                            media_repo[doc[field]] = S2InferenceError(f"Error processing {inferred_modality} file: {str(e)}")
+                    
+                    elif inferred_modality is Modality.TEXT and is_structured_index and media_field_types_mapping[field] in [FieldType.AudioPointer, FieldType.VideoPointer, FieldType.ImagePointer]:
+                        media_repo[doc[field]] = S2InferenceError(f"Error processing media file {doc}, detected as text or failed to determine modality")
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py]: @farshidz 
If we had a network issue when inferring the modality, it would not be included in the error message to the user. Working on this will require extra time. Is this generic error message acceptable?

File: src/marqo/tensor_search/add_docs.py (hunk 7 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: it might lead to an internal error which would fail the whole batch

catching it and passing to all the places which call infer_modality would take a bit of time. 

I think add_docs logic right now catches this failure but only returns something generic like 

`f"Error processing media file {doc}, detected as text or failed to determine modality"`

Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: There's surely a better way in the future though

File: src/marqo/tensor_search/add_docs.py [hunk 10]
```diff
@@ -88,17 +88,17 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
                 if field not in tensor_fields:
                     continue
                 if isinstance(doc[field], str) or force_download:
-                    modality = infer_modality(doc[field])
-                    if (modality == Modality.IMAGE and is_no_index) or (
-                        modality == Modality.IMAGE and is_unstructured_index) or (
-                        is_structured_index and media_field_types_mapping[field] == FieldType.ImagePointer):
+                    inferred_modality = infer_modality(doc[field])
+                    if (inferred_modality == Modality.IMAGE and is_no_index) or (
```
Comment (farshidz) [inline src/marqo/tensor_search/add_docs.py]: `is_no_index` is confusing. We shouldn't have code just to make testing easier

File: src/marqo/tensor_search/add_docs.py [hunk 11]
```diff
@@ -88,17 +88,17 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
...
+                        inferred_modality == Modality.IMAGE and is_unstructured_index) or (
+                        is_structured_index and media_field_types_mapping[field] == FieldType.ImagePointer and inferred_modality not in [Modality.TEXT, Modality.AUDIO, Modality.VIDEO]):
```
Comment (farshidz) [inline src/marqo/tensor_search/add_docs.py]: Why not in if image is what you're after?

File: src/marqo/tensor_search/add_docs.py [hunk 12]
```diff
@@ -127,27 +127,42 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
                                 else:
                                     raise e
 
-                    elif (modality in [Modality.VIDEO, Modality.AUDIO] and is_no_index) or (
-                            modality in [Modality.VIDEO, Modality.AUDIO] and is_unstructured_index) or (
-                            is_structured_index and media_field_types_mapping[field] in [FieldType.AudioPointer, FieldType.VideoPointer]):
-                            
+                    elif (inferred_modality in [Modality.VIDEO, Modality.AUDIO] and is_no_index) or (
+                            inferred_modality in [Modality.VIDEO, Modality.AUDIO] and is_unstructured_index) or (
+                            is_structured_index and media_field_types_mapping[field] in [FieldType.AudioPointer, FieldType.VideoPointer] and inferred_modality not in [Modality.TEXT, Modality.IMAGE]):
```
Comment (farshidz) [inline src/marqo/tensor_search/add_docs.py]: not in is still confusing and risky

**Review by farshidz — CHANGES_REQUESTED** [review]

**Comment by farshidz** [comment]
To mitigate regression risk, for structured image pointer we shouldn't use the infer process

File: src/marqo/s2_inference/multimodal_model_load.py [hunk 13]
```diff
@@ -143,27 +146,30 @@ def infer_modality(content: Union[str, List[str], bytes]) -> Modality:
...
+                with fetch_content_sample(encoded_url) as sample:
                     mime = magic.from_buffer(sample.read(), mime=True)
                     if mime.startswith('image/'):
                         return Modality.IMAGE
                     elif mime.startswith('video/'):
                         return Modality.VIDEO
                     elif mime.startswith('audio/'):
                         return Modality.AUDIO
-            except Exception as e:
+            except requests.exceptions.RequestException as e:
```
Comment (farshidz) [inline src/marqo/s2_inference/multimodal_model_load.py:173]: This should be an error for the document

File: src/marqo/tensor_search/add_docs.py (hunk 2 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: done

File: src/marqo/tensor_search/add_docs.py (hunk 11 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: removed this as discussed, infer_modality won't be used for structured image pointers

File: src/marqo/tensor_search/add_docs.py (hunk 12 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: changed to in [video, audio]

File: src/marqo/tensor_search/add_docs.py (hunk 9 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: Done, changed to pass the error to doc

File: src/marqo/tensor_search/add_docs.py (hunk 7 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: scratch the above. catching the error

File: src/marqo/tensor_search/add_docs.py (hunk 10 above)
Comment (RaynorChavez) [reply src/marqo/tensor_search/add_docs.py]: removed

File: src/marqo/s2_inference/multimodal_model_load.py [hunk 14]
```diff
@@ -143,28 +147,31 @@ def infer_modality(content: Union[str, List[str], bytes]) -> Modality:
...
                     if mime.startswith('image/'):
                         return Modality.IMAGE
                     elif mime.startswith('video/'):
                         return Modality.VIDEO
                     elif mime.startswith('audio/'):
                         return Modality.AUDIO
-            except Exception as e:
-                pass
+            except requests.exceptions.RequestException as e:
+                raise MediaDownloadError(f"Error downloading media file {content}: {e}")
```
Comment (wanliAlex) [inline src/marqo/s2_inference/multimodal_model_load.py]: raise MediaDownloadError from e

File: src/marqo/s2_inference/errors.py [hunk 15]
```diff
@@ -64,6 +66,8 @@ class BatchInferenceSizeNotMatchError(S2InferenceError):
 class ImageDownloadError(S2InferenceError):
     pass
 
+class MediaDownloadError(S2InferenceError):
```
Comment (wanliAlex) [inline src/marqo/s2_inference/errors.py:69]: Nice design!

File: src/marqo/s2_inference/multimodal_model_load.py [hunk 16]
```diff
@@ -143,28 +147,31 @@ def infer_modality(content: Union[str, List[str], bytes]) -> Modality:
...
                     mime = magic.from_buffer(sample.read(), mime=True)
                     if mime.startswith('image/'):
                         return Modality.IMAGE
                     elif mime.startswith('video/'):
                         return Modality.VIDEO
                     elif mime.startswith('audio/'):
                         return Modality.AUDIO
-            except Exception as e:
-                pass
+            except requests.exceptions.RequestException as e:
```
Comment (wanliAlex) [inline src/marqo/s2_inference/multimodal_model_load.py:173]: Do you think there will be other errors here? e.g., error from magic, mime, etc

**Review by wanliAlex — CHANGES_REQUESTED** [review]

File: src/marqo/s2_inference/multimodal_model_load.py (hunk 16 above)
Comment (RaynorChavez) [reply src/marqo/s2_inference/multimodal_model_load.py:173]: added

File: src/marqo/s2_inference/multimodal_model_load.py (hunk 14 above)
Comment (RaynorChavez) [reply src/marqo/s2_inference/multimodal_model_load.py]: done, but with slight modification

**Review by farshidz — DISMISSED** [review]

File: src/marqo/tensor_search/add_docs.py [hunk 17]
```diff
@@ -88,18 +90,23 @@ def threaded_download_and_preprocess_content(allocated_docs: List[dict],
...
-                    if (modality == Modality.IMAGE and is_no_index) or (
-                        modality == Modality.IMAGE and is_unstructured_index) or (
-                        is_structured_index and media_field_types_mapping[field] == FieldType.ImagePointer):
-
-                        if (marqo_index is not None
-                                and marqo_index.model.properties.get('type') in [ModelType.LanguageBind]
-                                and marqo_index.model.properties.get('supported_modalities') is not None
-                                and Modality.IMAGE not in marqo_index.model.properties.get('supported_modalities')):
+                    try:    
+                        inferred_modality = infer_modality(doc[field])
```
Comment (wanliAlex) [inline src/marqo/tensor_search/add_docs.py:94]: Doest this mean we are still inferring modality for the image field? We shouldn't do this to avoid regression.

**Review by wanliAlex — CHANGES_REQUESTED** [review]

**Review by wanliAlex — APPROVED** [review]
